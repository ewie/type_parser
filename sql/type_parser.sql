CREATE EXTENSION type_parser;

-- test builtin types
WITH t (type_str) AS (
  VALUES
    ('bit varying (42)'),
    ('bool'),
    ('boolean'),
    ('bpchar (7)'),
    ('char (42)'),
    ('character (37)'),
    ('character varying (37)'),
    ('date'),
    ('decimal (6,3)'),
    ('decimal'),
    ('double precision'),
    ('float'),
    ('float4'),
    ('int'),
    ('int4'),
    ('integer'),
    ('interval day to hour'),
    ('interval day to minute'),
    ('interval day to second (3)'),
    ('interval day to second'),
    ('interval day'),
    ('interval hour to minute'),
    ('interval hour to second (3)'),
    ('interval hour to second'),
    ('interval hour'),
    ('interval minute to second (3)'),
    ('interval minute to second'),
    ('interval month'),
    ('interval second (3)'),
    ('interval second'),
    ('interval year to month'),
    ('interval year'),
    ('interval year'),
    ('interval'),
    ('numeric (8,2)'),
    ('numeric'),
    ('real'),
    ('text'),
    ('text[][]'),
    ('time (3) with time zone'),
    ('time with time zone'),
    ('time without time zone'),
    ('time'),
    ('timestamp (3) with time zone'),
    ('timestamp (3) without time zone'),
    ('timestamp'),
    ('timestamptz (3)'),
    ('timetz (3)'),
    ('timetz'),
    ('varbit (23)'),
    ('varchar (42)')
)
SELECT
  t.type_str,
  format_type(p.typid, p.typmod)
FROM t
  CROSS JOIN parse_type_string(type_str) p
ORDER BY 1, 2;

-- test NULL as type string
\pset null '<NULL>'
SELECT parse_type_string(NULL);

-- test nonexistent type
SELECT parse_type_string('no_such_type');

-- test custom type
CREATE TYPE mytype AS (a int);
SELECT format_type(typid, typmod)
FROM parse_type_string('mytype');

-- test shell type
CREATE TYPE myshell;
SELECT parse_type_string('myshell');
