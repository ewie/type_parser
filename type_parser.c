#include "postgres.h"
#include "fmgr.h"
#include "funcapi.h"            /* for returning composite type */
#include "utils/builtins.h"     /* text_to_cstring() */
#include "parser/parse_type.h"  /* parseTypeString() */

PG_MODULE_MAGIC;

Datum parse_type_string(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(parse_type_string);

Datum
parse_type_string(PG_FUNCTION_ARGS)
{
#define PARSE_TYPE_STRING_COLS 2
  const char *type;    /* the type string we want to resolve */
  Oid         typid;   /* the resolved type oid */
  int32       typmod;  /* the resolved type modifier */
  TupleDesc   tupdesc;
  HeapTuple   rettuple;
  Datum       values[PARSE_TYPE_STRING_COLS] = {0};
  bool        nulls[PARSE_TYPE_STRING_COLS] = {0};

  type = text_to_cstring(PG_GETARG_TEXT_PP(0));

  /* Build a tuple descriptor for our result type */
  if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE) {
    ereport(ERROR,
            (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
             errmsg("function returning record called in context that cannot accept type record")));
  }

  BlessTupleDesc(tupdesc);

  /* Resolve type oid and modifier. */
  parseTypeString(type, &typid, &typmod, fcinfo->context);

  /* Create return tuple. */
  values[0] = typid;
  values[1] = typmod;
  rettuple = heap_form_tuple(tupdesc, values, nulls);

  return HeapTupleGetDatum(rettuple);
#undef PARSE_TYPE_STRING_COLS
}
