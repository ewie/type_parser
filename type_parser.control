default_version = '0.1'
module_pathname = '$libdir/type_parser'
relocatable = false
