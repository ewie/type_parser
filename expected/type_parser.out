CREATE EXTENSION type_parser;
-- test builtin types
WITH t (type_str) AS (
  VALUES
    ('bit varying (42)'),
    ('bool'),
    ('boolean'),
    ('bpchar (7)'),
    ('char (42)'),
    ('character (37)'),
    ('character varying (37)'),
    ('date'),
    ('decimal (6,3)'),
    ('decimal'),
    ('double precision'),
    ('float'),
    ('float4'),
    ('int'),
    ('int4'),
    ('integer'),
    ('interval day to hour'),
    ('interval day to minute'),
    ('interval day to second (3)'),
    ('interval day to second'),
    ('interval day'),
    ('interval hour to minute'),
    ('interval hour to second (3)'),
    ('interval hour to second'),
    ('interval hour'),
    ('interval minute to second (3)'),
    ('interval minute to second'),
    ('interval month'),
    ('interval second (3)'),
    ('interval second'),
    ('interval year to month'),
    ('interval year'),
    ('interval year'),
    ('interval'),
    ('numeric (8,2)'),
    ('numeric'),
    ('real'),
    ('text'),
    ('text[][]'),
    ('time (3) with time zone'),
    ('time with time zone'),
    ('time without time zone'),
    ('time'),
    ('timestamp (3) with time zone'),
    ('timestamp (3) without time zone'),
    ('timestamp'),
    ('timestamptz (3)'),
    ('timetz (3)'),
    ('timetz'),
    ('varbit (23)'),
    ('varchar (42)')
)
SELECT
  t.type_str,
  format_type(p.typid, p.typmod)
FROM t
  CROSS JOIN parse_type_string(type_str) p
ORDER BY 1, 2;
            type_str             |          format_type           
---------------------------------+--------------------------------
 bit varying (42)                | bit varying(42)
 bool                            | boolean
 boolean                         | boolean
 bpchar (7)                      | character(7)
 char (42)                       | character(42)
 character (37)                  | character(37)
 character varying (37)          | character varying(37)
 date                            | date
 decimal                         | numeric
 decimal (6,3)                   | numeric(6,3)
 double precision                | double precision
 float                           | double precision
 float4                          | real
 int                             | integer
 int4                            | integer
 integer                         | integer
 interval                        | interval
 interval day                    | interval day
 interval day to hour            | interval day to hour
 interval day to minute          | interval day to minute
 interval day to second          | interval day to second
 interval day to second (3)      | interval day to second(3)
 interval hour                   | interval hour
 interval hour to minute         | interval hour to minute
 interval hour to second         | interval hour to second
 interval hour to second (3)     | interval hour to second(3)
 interval minute to second       | interval minute to second
 interval minute to second (3)   | interval minute to second(3)
 interval month                  | interval month
 interval second                 | interval second
 interval second (3)             | interval second(3)
 interval year                   | interval year
 interval year                   | interval year
 interval year to month          | interval year to month
 numeric                         | numeric
 numeric (8,2)                   | numeric(8,2)
 real                            | real
 text                            | text
 text[][]                        | text[]
 time                            | time without time zone
 time (3) with time zone         | time(3) with time zone
 timestamp                       | timestamp without time zone
 timestamp (3) without time zone | timestamp(3) without time zone
 timestamp (3) with time zone    | timestamp(3) with time zone
 timestamptz (3)                 | timestamp(3) with time zone
 timetz                          | time with time zone
 timetz (3)                      | time(3) with time zone
 time without time zone          | time without time zone
 time with time zone             | time with time zone
 varbit (23)                     | bit varying(23)
 varchar (42)                    | character varying(42)
(51 rows)

-- test NULL as type string
\pset null '<NULL>'
SELECT parse_type_string(NULL);
 parse_type_string 
-------------------
 <NULL>
(1 row)

-- test nonexistent type
SELECT parse_type_string('no_such_type');
ERROR:  type "no_such_type" does not exist
